### PR TITLE
Widget Testing Harness

### PR DESCRIPTION
**Pull request by kitsonk**
**Type:** feature

The following has been addressed in the PR:

* [X] There is a related issue
* [X] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [X] Unit or Functional tests are included in the PR

**Description:**

This PR provides several new modules that make testing widgets easier and more complete, without the need to "break" the visibility of a widget class.

The main usage is to "harness" a widget class and then perform tests/checks against that harnessed widget class.  The harness will render the widget to the DOM using a stripped down virtual DOM projector, which injects certain "spies" to facilitate testing.  The developer interacts with the widget class in a similar way it would utilised in a larger application.  Because the harness will render the widget to the DOM, the developer has the ability to not only introspect the virtual DOM generated, but can validated that that virtual DOM materialises itself as expected in the real DOM.

A basic usage of testing the widget using Intern would look something like this:

```typescript
import * as registerSuite from 'intern!object';
import * as assert from 'intern/chai!assert';
import harness from '@dojo/intern-helper/harness';
import { v, w } from '@dojo/widget-core/d';

import MyWidget from '../../MyWidget';
import css from '../../styles/myWidget.m.css';

registerSuite({
    name: 'MyWidget',

    basic() {
        const widget = harness(MyWidget);
        widget.expectRender(v('div', { classes: widget.classes(css.root) }, [ w('child', {
            bind: true,
            key: 'first-child',
            classes: widget.classes(css.child)
        }, 'should render as expected');

        widget.destroy();
    }
});
```

There are also several class aligned APIs that are available on a `harness`ed class:

- `.listener` - a reference to a simple stub function to use on an expected render to use as a place holder for listener functions in a render.
- `.classes()` - returns a value to be passed as a `classes` property of a `v()` or `w()` call that adds classes in the same way they would be added to a virtual DOM node.
- `.expectedRender()` - provide an expected of virtual DOM which will be compared with the actual rendered virtual DOM from the widget class.
- `.getRender()` - Provide the Dojo virtual DOM for advanced introspection.  It is important to note though that there is some post processing done on the virtual DOM by this point via the `.__render__()` method on the harnessed class.
- `.getDom()` - Return the root node of the render DOM.
- `.resetClasses()` - Reset the classes to "forget" any previously asserted classes.
- `.sendEvent()` - Dispatch an event to the DOM of the rendered widget.  It accepts a map of options which allow configuring exactly what class of event, properties and allows sub-selection of the rendered DOM.
- `.setChildren()` - Provide children that should be passed to the widget class as it is rendered.  These are typically passed by an upstream widget by invoking a `w(WidgetClass, { }, [ ...children ])`.
- `.setProperties()` - Provide a map of properties to a widget.  These are typically passed by an upstream widget by invoking a `w(WidgetClass, { ...properties })`.

Resolves dojo/widgets#66

### CODE REVIEW

**Review by matt-gadd — APPROVED** [review]

File: src/harness.ts [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+}
+
+/**
+ * A private class that is used to actually render the widget and keep track of the last render by
+ * the harnessed widget.
+ */
+class WidgetHarness<P extends WidgetProperties, W extends typeof WidgetBase> extends WidgetBase<P> {
+	private _widgetConstructor: W;
+	private _afterCreate: (element: HTMLElement) => void;
+	private _id: string = ROOT_CUSTOM_ELEMENT_NAME + '-' + (++harnessId);
```
Comment (agubler) [inline src/harness.ts]: Does this cause a lint failure in tslint 4+? that you shouldn't need to define the type `string` (i think)

Comment (kitsonk) [reply src/harness.ts]: Hmmm... for some reason, the trivially inferred type was not flagged.  I wonder if `tslint` didn't detect the implicit coercion, though TypeScript does.  Anyways, changed.

File: src/harness.ts [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+		return this;
+	}
+}
+
+/**
+ * Harness a widget class for testing purposes, returning an API to interact with the harness widget class.
+ * @param widgetConstructor The constructor function/class of widget that should be harnessed.
+ * @param projectionRoot The root where the harness should append itself to the DOM.  Default to `document.body`
+ */
+export default function harness<P extends WidgetProperties, W extends typeof WidgetBase>(widgetConstructor: W, projectionRoot?: HTMLElement): Harness<P, W> {
```
Comment (agubler) [inline src/harness.ts]: Can this be? 
```ts
export default function harness<P extends WidgetProperties, W extends Constructor<WidgetBase<P>>>(widgetConstructor: W, projectionRoot?: HTMLElement): Harness<P, W> {
```

File: src/harness.ts [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+		const { target = this.getDom() } = options;
+		sendEvent(target, type, options);
+		return this;
+	}
+
+	/**
+	 * Set the children that will be used when rendering the harnessed widget
+	 * @param children The children to be set on the harnessed widget
+	 */
+	public setChildren(...children: DNode[]): this {
```
Comment (agubler) [inline src/harness.ts]: Should this be the same API as `WidgetBase` so `public setChildren(children: DNode[])...`?

Comment (kitsonk) [reply src/harness.ts]: We sort of discussed this before... `.setChildren(...children)` is in my opinion more "usable" and we discussed that `WidgetBase.setChildren()` isn't actually meant to be a consumable API (and that we were considering renaming it `.__setChildren__()` to indicated that).  Given that people shouldn't be using it anyways, there is no need for them to conform, and if they don't need to conform, why not use the more usable API?  Is there really any value in a single array argument?  If you have an array of children, just spread them.  If not, you can avoid the literal array braces.

Comment (agubler) [reply src/harness.ts]: > If you have an array of children, just spread them. If not, you can avoid the literal array braces.

Conversely if you don't have an array just wrap it in `[]` and if you do have an array just use it. I just think it's 6 of one, half dozen of the other but my suggestion was more for consistency - Indeed `setChildren` is not intended to be for public use with widgets and although we have talked about changing to `__setChildren__`, the `projector` would still need a `setChildren` API (and we haven't discussed changing the signature.)

Comment (kitsonk) [reply src/harness.ts]: It is a minor change, so I will change it.  I still feel that rest arguments are more usable.

**Review by agubler — APPROVED** [review]

**Pull request merged**